internal/service/agent/tasks.go

```diff
@@ -25,39 +34,45 @@ func UpdateMetrics(cache storage.AgentCache, monitor *monitors.Monitor, rand *ut
 	}
 }
 
-func SendMetrics(client *resty.Client, url string, endpoint string, cache storage.AgentCache, logger *zap.Logger) {
+func SendMetrics(client *resty.Client, url string, endpoint string, cache cache.AgentCache, logger *zap.Logger) {
 	gauges, counters := cache.GetMetrics()
 
 	err := cache.ResetMetricValue(string(metrics.Counter), "PollCount")
 	if err != nil {
 		logger.Error(err.Error(), zap.String("event", "reset PollCount counter metric"))
 	}
 
+	metricsBatch := make([]metrics.Metrics, 0)
+
 	for _, gaugeMetric := range gauges {
-		metric := metrics.GaugeMetricToMetrics(gaugeMetric)
-		body, err := json.Marshal(metric)
-		if err != nil {
-			logger.Error(err.Error(), zap.String("event", "JSON encoding gauge metric"))
-			continue
-		}
+		metricsBatch = append(metricsBatch, metrics.GaugeMetricToMetrics(gaugeMetric))
 
-		_, err = sendPostRequest(client, url, endpoint, "application/json", body)
-		if err != nil {
-			logger.Error(err.Error(), zap.String("event", "send update gauge metric"))
-		}
 	}
 
 	for _, counterMetric := range counters {
-		metric := metrics.CounterMetricToMetrics(counterMetric)
-		body, err := json.Marshal(metric)
-		if err != nil {
-			logger.Error(err.Error(), zap.String("event", "JSON encoding counter metric"))
-			continue
-		}
+		metricsBatch = append(metricsBatch, metrics.CounterMetricToMetrics(counterMetric))
 
-		_, err = sendPostRequest(client, url, endpoint, "application/json", body)
-		if err != nil {
-			logger.Error(err.Error(), zap.String("event", "send update counter metric"))
+	}
+
+	body, err := json.Marshal(metricsBatch)
+	if err != nil {
+		logger.Error(err.Error(), zap.String("event", "JSON encoding metrics batch"))
+	}
+
+	var sendErr error
+	for _, interval := range sendMetricsBatchRetryInterval {
```

Comment:
SUGGESTION: if I were you, I would use the `backoff` package.

https://pkg.go.dev/github.com/cenkalti/backoff/v4#Retry